tests/utils/validate_docstring_examples.py

```diff
@@ -0,0 +1,212 @@
+# License: MIT
+# Copyright © 2022 Frequenz Energy-as-a-Service GmbH
+
+"""Validate docstring code examples."""
+
+from __future__ import annotations
+
+import argparse
+import importlib.util
+import inspect
+import re
+import sys
+import textwrap
+import types
+from pathlib import Path
+
+
+def get_module_docs(module: types.ModuleType) -> list[str]:
+    """Get all docstrings found in the given module.
+
+    Args:
+        module: The module to extract docstrings from.
+
+    Returns:
+        A list of docstrings.
+    """
+    docs = []
+
+    for _, obj in inspect.getmembers(module):
+        if inspect.isfunction(obj) and obj.__module__ == module.__name__:
+            if obj.__doc__ is not None:
+                docs.append(obj.__doc__)
+        elif inspect.isclass(obj) and obj.__module__ == module.__name__:
+            if obj.__doc__ is not None:
+                docs.append(obj.__doc__)
+            for _, method_obj in inspect.getmembers(obj):
+                if (
+                    inspect.isfunction(method_obj)
+                    and method_obj.__module__ == module.__name__
+                ):
+                    if method_obj.__doc__ is not None:
+                        docs.append(method_obj.__doc__)
+
+    return docs
+
+
+def extract_code_examples(module: types.ModuleType) -> list[str]:
+    """Extract all code examples from the docstrings found in a module.
+
+    Args:
+        module: The module to extract code examples from.
+
+    Returns:
+        A list of code examples.
+    """
+    code_blocks = []
+    for doc in get_module_docs(module):
+        code_blocks += re.findall(r"``` ?python(.*?)```", doc, re.DOTALL)
+
+    return code_blocks
+
+
+# pylint: disable=too-many-locals
+def run_code_examples(
```

Comment:
Also in the future I would just write `pytest` files, so we can use the full power of pytest when running tests (no need to implement all the options here, like `fail_fast` or `verbose`) and we could also [mark tests](https://docs.pytest.org/en/7.1.x/how-to/mark.html#registering-marks) like "skip", or "release" to run some tests on releases only.

Reply:
I did try to use it first but eventually didn't explore that branch fully. I can have a look again